lib/config/dynamic/admiral/source.go

```diff
@@ -0,0 +1,497 @@
+// Get returns the dynamic config portion from an Admiral instance. For now,
+// this is empty pending details from the Admiral team.
+func (a *source) Get(ctx context.Context) (*vchcfg.VirtualContainerHostConfigSpec, error) {
+	a.mu.Lock()
+	lastCfg := a.lastCfg
+	a.mu.Unlock()
+
+	c, projs, err := a.discover(ctx)
+	if err != nil {
+		log.Warnf("error locating Admiral, returning last known config: %s", err)
+		return lastCfg, nil
+	}
+
+	if len(projs) == 0 {
+		return nil, nil
+	}
+
+	wl, err := a.whitelist(ctx, c, projs)
+	if err != nil {
+		log.Warnf("error getting whitelist form Admiral, returning last known config: %s")
+		return lastCfg, nil
+	}
+
+	log.Debugf("got whitelist: %+v", wl)
+
+	newCfg := &vchcfg.VirtualContainerHostConfigSpec{
+		Registry: vchcfg.Registry{RegistryWhitelist: wl},
+	}
+
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	err = nil
+	if !a.diff(newCfg) {
+		err = dynamic.ErrConfigNotModified
+	}
+
+	a.lastCfg = newCfg
+	return newCfg, err
+}
+
+func (a *source) diff(newCfg *vchcfg.VirtualContainerHostConfigSpec) bool {
+	if a.lastCfg == nil {
+		if newCfg == nil {
+			return false
+		}
+
+		return true
+	}
+
+	if newCfg == nil {
+		return true
+	}
+
+	// compare whitelists
+	if len(a.lastCfg.RegistryWhitelist) != len(newCfg.RegistryWhitelist) {
+		return true
+	}
+
+	for _, w1 := range a.lastCfg.RegistryWhitelist {
+		found := false
+		for _, w2 := range newCfg.RegistryWhitelist {
+			if w2 == w1 {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (a *source) discover(ctx context.Context) (*client.Admiral, []string, error) {
+	a.mu.Lock()
+	c := a.c
+	v := a.v
+	a.mu.Unlock()
+
+	var vms []*vm.VirtualMachine
+	removed := true
+	if c != nil {
+		projs, err := a.projects(ctx, c)
+		if err != nil {
+			log.Debugf("error getting projects: %s", err)
+			vms = append(vms, v)
+			removed = false
+		} else if len(projs) > 0 {
+			return c, projs, nil
+		}
+	}
+
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	if c != a.c && a.c != nil {
+		return a.c, a.projs, nil
+	}
+
+	var err error
+	// if there isn't a list of potential product VMs
+	// already, then run discovery
+	if len(vms) == 0 {
+		log.Infof("running product VM discovery")
+		vms, err = a.d.Discover(ctx, a.sess)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	for _, v := range vms {
+		u, token, err := admiralEndpoint(ctx, v)
+		if err != nil {
+			log.Warnf("ignoring potential product VM: %s", err)
+			continue
+		}
+
+		rt := rtclient.NewWithClient(u.Host, u.Path, []string{u.Scheme}, admClient)
+		rt.DefaultAuthentication = &admiralAuth{token: token}
+		c := client.New(rt, strfmt.Default)
+		projs, err := a.projects(ctx, c)
+		if err == nil && len(projs) > 0 {
+			log.Infof("using Admiral endpoint at %s", u)
+			a.c = c
+			a.projs = projs
+			a.v = v
+			return c, projs, nil
+		}
+
+		log.Warnf("ignoring product VM Admiral endpoint %s since it does not have the VCH added to any project", u)
+	}
+
+	if removed {
```

Comment:
Should we move this before the Discover? do we need a full cycle of "removed" before we check to see if there's a different Admiral that claims this VCH.